Content.Shared/Chemistry/Components/Solution.cs

```diff
@@ -674,58 +675,97 @@ public Solution SplitSolution(FixedPoint2 toTake)
             return newSolution;
         }
 
+        // Frontier: cryogenics per-reagent filter function (#1443)
         /// <summary>
-        /// Frontier
-        /// splits the solution taking up to the specified amount of each reagent from the solution.
-        /// If the solution has less of a reagent than the specified amount, it will take all of that reagent.
+        /// Splits a solution, taking the specified amount of each reagent from the solution.
+        /// If any reagent in the solution has less volume than specified, it will all be transferred into the new solution.
         /// </summary>
-        /// <param name="toTakePer">How much of each reagent to take</param>
-        /// <returns>a new solution containing the reagents taken from the original solution</returns>
-        public Solution SplitSolutionReagentsEvenly(FixedPoint2 toTakePer)
+        /// <param name="toTakePer">How much of each reagent to take.</param>
+        /// <returns>A new solution containing the reagents taken from the original solution.</returns>
+        public Solution SplitSolutionPerReagent(FixedPoint2 toTakePer)
         {
-            var splitSolution = new Solution();
-
             if (toTakePer <= FixedPoint2.Zero)
-                return splitSolution;
-            var reagentsCount = Contents.Count;
-            var reagentsToRemove = new List<ReagentQuantity>();
-            for (var i = 0; i < reagentsCount; i++)
-            {
-                var currentReagent = Contents[i];
+                return new Solution();
 
-                if (currentReagent.Quantity <= FixedPoint2.Zero)
-                {
-                    reagentsToRemove.Add(currentReagent);
-                    continue;
-                }
+            var origVol = Volume;
+            Solution newSolution = new Solution(Contents.Count) { Temperature = Temperature };
+
+            for (var i = Contents.Count - 1; i >= 0; i--) // iterate backwards because of remove swap.
+            {
+                var (reagent, quantity) = Contents[i];
 
-                if (currentReagent.Quantity <= toTakePer)
+                // If the reagent has more than enough volume to remove, no need to remove it from the list.
+                if (quantity > toTakePer)
                 {
-                    splitSolution.AddReagent(currentReagent);
-                    reagentsToRemove.Add(currentReagent);
+                    Contents[i] = new ReagentQuantity(reagent, quantity - toTakePer);
+                    newSolution.Contents.Add(new ReagentQuantity(reagent, toTakePer));
+                    Volume -= toTakePer;
                 }
                 else
                 {
-                    splitSolution.AddReagent(currentReagent.Reagent, toTakePer);
-                    RemoveReagent(currentReagent.Reagent, toTakePer);
+                    Contents.RemoveSwap(i);
```

Comment:
as I had mentioned in the private messages, it is bad practice to remove from an iterator while iterating threw it.

the argument of "but everywhere else does it" falls in line with "if they jump off a bridge would you"
quality can not be improved if it is ignored for consistently
that's not directed at you, but anybody and everybody who works on code
as developers it is our responsibility to protect quality where we can